3scale/porta
# Segment REST + Bugfix of number of requests

### What this PR means to solve
1. We used to use the `GraphQL` API to delete users in Segment because that was the only API available for this matter back then. But now that API is deprecated and not working anymore, and this PR moves to the `REST` API. Closes [THREESCALE-3395 - Segment GraphQL deprecation on 2019, September 10th](https://issues.jboss.org/browse/THREESCALE-3395)
2. We used to make a request for each user destroyed to Segment right away, but we used to have problems of `rate limit`, so this PR also aims to fix that. Closes [THREESCALE-2840 - Segment::ClientError in SegmentDeleteUserWorker@default](https://issues.jboss.org/browse/THREESCALE-2840)


### Tasks
- [X] The 1st commit only removes all the previous implementation of removing users from Segment. It only does not remove the code and test to fetch [the config file](https://github.com/3scale/deploy/pull/610).
- [X] When a user is deleted, it is saved in DeletedObject.
- [X] The job config runs a worker daily that deletes from segment those users that have been deleted previously.
- [X] Remove from "DeletedObject" those users that have already been deleted in Segment. This way those users won't be tried to be removed the next day again. And because we don't need them there anymore at that point.
- [X] Remove all the previously deleted users from Segment. For the time when the GraphQL has not worked and we have been destroying users in SaaS anyway, so now we will need to use this new implementation to remove also those. The plan is to have a csv with all the users we have in Segment, and run a rake task that for each one of them, if it is not in our DB as a user, we save it as a deleted object. This way the job cron will remove them from Segment as well later on.

### Verification steps
I've also tested it in preview, but just to know if it works, it can be done locally as explained here.
1. Check a few user ids that we actually want to remove for real. Do it by doing this in the rails console of Preview: `Account.deleted_since.first(3).map { |account| account.users.pluck(:id) }.flatten`. For the next steps, let's imagine that those user ids are these: `[-1, -2, -3]`. I wrote them in negative on purpose just in case that someone does not see that it is just an example and actually executes, so we won't be removing real data of real non-deleted users by accident.
2. Locally, in the development environment, in the rails console, create the users with those ids and destroy them to create the DeletedObject by the code.
3. Copy the config of Segment from `config/examples/features.yml` to `config/features.yml`. Use the correct configuration credentials from [the deploy repository](https://github.com/3scale/deploy/pull/610). Do not forget to enable the config.
4. Run the low queue of Sidekiq. `bundle exec sidekiq -q low -L log/sidekiq.log`
5. In the rails console opened after the config is changed, run the worker. `SegmentDeleteUsersWorker.perform_now`
6. See the [regulations created](https://app.segment.com/3scale/settings/deletion-requests) as expected. You can also do locally from your terminal: `curl "https://platform.segmentapis.com/v1beta/workspaces/{workspace_name}/regulations" --header "Authorization: Bearer YOUR_TOKEN"`
